url/urlsearchparams.go

```diff
@@ -0,0 +1,410 @@
+package url
+
+import (
+	"fmt"
+	"net/url"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/dop251/goja"
+)
+
+var (
+	reflectTypeString = reflect.TypeOf("")
+	reflectTypeObject = reflect.TypeOf(map[string]interface{}{})
+	reflectTypeArray  = reflect.TypeOf([]interface{}{})
+	reflectTypeMap    = reflect.TypeOf([][2]interface{}{})
+)
+
+func newInvalidTupleError(r *goja.Runtime) *goja.Object {
+	return newError(r, "ERR_INVALID_TUPLE", "Each query pair must be an iterable [name, value] tuple")
+}
+
+func newMissingArgsError(r *goja.Runtime, msg string) *goja.Object {
+	return newError(r, "ERR_MISSING_ARGS", msg)
+}
+
+func newInvalidArgsError(r *goja.Runtime) *goja.Object {
+	return newError(r, "ERR_INVALID_ARG_TYPE", `The "callback" argument must be of type function. Received undefined`)
+}
+
+func newError(r *goja.Runtime, code string, msg string) *goja.Object {
+	o := r.NewTypeError("[" + code + "]: " + msg)
+	o.Set("code", r.ToValue(code))
+	return o
+}
+
+// Currently not supporting the following:
+//   - ctor(iterable): Using function generators
```

Comment:
Is this still relevant?

Reply:
With the current implementation this is still relevant. I don't have any code to handle the use of generator functions. It's treated as an Object that generates an empty object. I would need to find a way to call the generator function and process it's output. Any leads on how I could do that given a `goja.Object` that represents a generator function?

Reply:
The output of a generator function is an iterable for which you can use ForOf. BTW, an array is also an iterable, so you can generalise it for objects:

```go
func buildParamsFromObject(r *goja.Runtime, v goja.Value) *url.URL {
    o := v.ToObject(r)
    if o.GetSymbol(goja.SymIterator) != nil {
        return buildParamsFromIterable(r, o)
    }
    ...
}
```
`buildParamsFromIterable` is pretty much the same as the current `buildParamsFromArray`.

Reply:
Awesome! I totally missed the `goja.GetSymbol()` functionality in goja. This is great! I'll make this change and push it up. Thanks for the heads up.